text/0000-codec-identifiers.md

```diff
@@ -0,0 +1,94 @@
+
+### Proposed API
+
+Codec tags can change depending on encoder configuration, and cannot be a static value in the `obs_encoder_info` struct, thus there needs to be a function that can be called to ask the encoder for the appropiate tag for its current configuration.
+
+- New API function: `uint32_t obs_encoder_get_codec_tag(obs_encoder_t *encoder)`
+- New `obs_encoder_info` struct member: `uint32_t (*codec_tag)(void *data)`
+
+For most codecs the implementation of the `codec_tag` function may be ommited and a default value is provided by libobs. This default tag could be obtained from libavcodec via `av_codec_get_tag()` or hardcoded to the ones we currently use (i.e. `avc1` for H.264, `hvc1` for HEVC, `av01` for AV1).
+
+**Note:** FFmpeg currently has two separate codec tag lists, one for ISO-BMFF/QuickTime (MP4/MOV) and one for RIFF (e.g. AVI), this proposal would only implement the MOV/MP4 tags as the RIFF ones are not relevant to OBS outside of possibly DirectShow.
+We may elect to remove ambiguities by having a second paramter specifying a codec tag namespace, or explicitly naming the function to indicate that it returns ISO format tags.
+
+## Codec Identifier
+
+This RFC proposes there different methods, one of which should be agreed upon:
+
+### FFmpeg AVCodecID
+
+This approach would simply use `AVCodecID` enum defined in libavcodec.
+
+#### Advantages
+
+- libobs already links against libav* libraries so IDs can be used without many changes
+- Easy to use in FFmpeg muxer and FFmpeg-based outputs (e.g. SRT/RIST)
+- libav* functions available for converting from current string-based names to FFmpeg IDs
+
+#### Drawbacks
+
+- Not all modules that interface with video data or implement encoders currently depend on ffmpeg (e.g. obs-outputs, VideoToolbox)
+- While unlikely, it could happen that FFmpeg does not support a codec we want to support, or support is missing from an older version, e.g. on LTS releases
+
+### Codec Tags / FourCC
+
+In this case we define the aforementioned codec tag as the new identifier.
+
+#### Advantages
+
+- Avoids a secondary API for codec identifiers
+- FFmpeg's ISO/QuickTime tag list should support all codecs upstream OBS supports
+
+#### Drawbacks
+
+- Many codecs have more than one tag, which could require large if clauses and may still require mapping to an internal enumeration value
+- Codec support for ISO and QuickTime formats is limited
+- Tags are mostly intended to be opaque values that are relevant to the muxer, and probably shouldn't be used as a unqiue identifier for a codec family
+
+### OBS-specific IDs
+
+In this case we would define our own enumeration, for example:
+```c
+enum OBSCodecID {
+    OBS_CODEC_INVALID = 0,
+    OBS_CODEC_H264,
+    OBS_CODEC_HEVC,
+    OBS_CODEC_AV1,
+    OBS_CODEC_PRORES,
+}
+```
+
+#### Advantages
+
+- Simple enum in libobs
+- No external dependencies
+
+#### Drawbacks
+
+- Requires libobs API changes to add new codecs
```

Comment:
Wouldn't all three solutions require libobs API changes?

Reply:
The 3 solution requires to modify the API at **each codec addition** unlike the two others that relies on a third-party enum or FourCC (not an enum but integers).